vpr/src/route/connection_router.h

```diff
@@ -234,6 +240,10 @@ class ConnectionRouter : public ConnectionRouterInterface {
     RouterStats* router_stats_;
     HeapImplementation heap_;
     bool router_debug_;
+
+    public:
+    // The path manager for RCV, keeps track of the route tree as a set, also manages the allocation of the heap types
+    PathManager rcv_path_manager;
```

Comment:
This should be a private variable, and it likely should be a pointer not a copy?

Reply:
I guess a better question is why you left this as a public member variable?

Reply:
I knew you wouldn't like this part! It's only public because it needs to be accessed by update_route_tree function, which isn't a member of connection router. I could make it private and have an interface in connection router?

Reply:
That's a really bad reason to do this, best to trust your instinct here.  It looks like the path manager needs the set of visited nodes during `update_route_tree`, but doesn't actually need to be updated in real time anyways.  It is better to do a bulk update of nodes_visited rather than incrementally during `update_route_tree`, so find a way to do that.  Also the current code structure means that if not running in RCV you are taking a penalty.

Reply:
```
src/route/route_path_manager.cpp
41:void PathManager::insert_node(RRNodeId node) {

src/route/route_tree_timing.cpp
332:        rcv_path_manager.insert_node(RRNodeId(inode));
396:        rcv_path_manager.insert_node(RRNodeId(rr_node));

src/route/route_path_manager.h
43:    void insert_node(RRNodeId node);
```

Looks like node insertion is **only** in route tree timing, so it doesn't seem like `route_tree_nodes` belongs in `PathManager` at all.  If you look at the access pattern of `route_tree_nodes`, `PathManager` simply exposes to access `route_tree_nodes` directly, and never actually manipulates it.  This is a pretty strong sign it doesn't belong.

Reply:
Yeah insertion only happens in update_route_tree. The only reason route_tree_nodes is required, is for the node_exists_in_tree function also in the PathManager class. Due to this, I think it's inclusion in PathManager is beneficial, as it cleans up the connection router code, and isolates the management to the PathManager class. 

My older solution was to pass route_tree_nodes as an argument throughout connection_router, which has no benefit over this solution.

As for updating it in real time, I don't see the benefit to doing it in bulk, over doing it during the route tree updating. The current code structure means that if RCV is disabled, the is_enabled flag (to be renamed) will be disabled and thus there will be no runtime impact.

Reply:
This has been fixed as of one of my newer commits.